### ISSUE TITLE
Bootstrapping (respecting SAP license terms)

### ISSUE THREAD

**Issue by marianfoo** [oblomov-dev/abap2UI5#123]
I suspect abap2ui5 is currently used mainly in the on-premise context.
In the current configuration SAPUI5 is used by the CDN ui5.sap.com.
Unfortunately this is not allowed see SAP Note [2943781](https://launchpad.support.sap.com/#/notes/2943781).

> The usage of the SAPUI5 CDN is only allowed for SAP Cloud products like e.g. SAP Business Technology Platform, SAP S/4HANA Cloud, public edition or SAP SuccessFactors, however on premise installations/systems are not allowed to bootstrap SAPUI5 from the SAPUI5 CDN.

I recommend making the `repository` parameter configurable and setting it during setup.
https://github.com/oblomov-dev/abap2UI5/blob/03856e3d0c51a7d8c35e4c4cc4a2037dbefc15c0/src/z2ui5_cl_http_handler.clas.abap#L12
So everyone can set his own UI5 variant for which licenses were purchased.
Possible settings would be e.g. src="/sap/public/bc/ui5_ui5/resources/sap-ui-core.js" , src="resources/sap-ui-core.js" , an own CDN or even OpenUI5, see note [3207822](https://launchpad.support.sap.com/#/notes/3207822).

**Comment by larshp** [comment]
heh, _insert rant about SAP here_

In some contexts SAP thinks that applications built by customers are SAP Products, because SAP doesnt want that setup to break license terms. In some contexts SAP argues that even tho something is breaking license terms, its not their intention to enforce the license terms.

🤷‍♂️

**Comment by marianfoo** [comment]
I know it is a community open source project that is also promoted by the SAP Developer Advocacy team.
I don't think it will be enforced, especially here.

But the Notes clearly state that the CDN may not be used for on-premise and that this may cause big problems, be it only the internal compliance department.

I would definitely remove the hardcoded CDN and describe that the CDN can be used in certain scenarios with the note.
Then everyone can decide for themselves whether they have the licence for it.

**Comment by oblomov-dev** [comment]
Ah thanks for the hint! Did not know this and never expected 😯 
SAP makes it complicated sometimes....

I added now this and will also update the documentation (so everyone can decide on its own):
<img width="795" alt="image" src="https://user-images.githubusercontent.com/102328295/227510024-86ff28e2-99a1-4981-b9dc-6b413c9e1603.png">

@marianfoo 
What do you mean in detail with „making the repository parameter configurable and setting it during setup“ ?

For me a constant is the idea to make it configurable and I don’t want to have a customizing table, want to stay source code only, but let me know your ideas.

**Comment by marianfoo** [comment]
Personally, I would not use this line and let the developer define this explicitly  in the source code itself and not use the CDN as a default option.
https://github.com/oblomov-dev/abap2UI5/blob/03856e3d0c51a7d8c35e4c4cc4a2037dbefc15c0/src/z2ui5_cl_http_handler.clas.abap#L12
you know your code better but something like this as a prerequisite

```ABAP
z2ui5_if_app->set_ui5_src( 'resources/sap-ui-core-js' )
````

**Comment by margopolo** [comment]
I support the proposal by @marianfoo. Even if you ignore the license issue it makes sense to have some kind of configuration for bootstrapping UI5. In your current implementation you reference to the default SAPUI5 version, but this URL must only be used for testing purpose and not for productive usage as "_The default version is constantly being upgraded and this might have an impact on the stability of your application._" (see also the [SAPUI5 documentation](https://ui5.sap.com/#/topic/2d3eb2f322ea4a82983c1c62a33ec4ae)). So also in case of using CDN consumers should be able to specify/configure the SAPUI5 version the want/need to use.
I could imagine this (configuration) is also a valid point for the theme - you never know when a new theme is coming or rather quite some customers are still prefer to use sap_fiori_3 ... (but guess this is a separate issue).

**Comment by oblomov-dev** [comment]
@larshp @marianfoo @margopolo thanks for the infos! Few thoughts:

App Configuration (Theme/Repository...):
- configuring this via interface is also for me the right way. In normal UI5 app development you also define it with the index.html for every app, so it would make sense to make it also a part of Z2UI_IF_APP. But I am not sure if abap2UI5 works like this. My idea is that an app here is a smaller unit, I prepared some demos (popup, flow logic) where you can see that you can switch in abap2UI5 between apps easily like when you use a function module to display a popup and then go back to the original app. I think I have to add this therefore on a higher level, but this will be a bigger change.
- Same with the theme, I could add this to Z2UI5_IF_APP, but i am also not sure if it is the right place (a popup then could have an other theme than the main app) or better add it on a higher more general level.
- So far both is configurable via constants in Z2UI5_CL_HTTP_HANDLER, it is in the z namespace so every customer can change it. I will think about a better place for this, maybe adding a new interface for general configuration.

Bootstrapping:
I have in mind that abap2UI5 is based on (1) one source line, (2) running everywhere, (3) out of the box. Means install it with abapGit and everything is ready to start without additional configuration and without the need to take care if you are „abap cloud“ or “standard abap“. 
Same for apps: develop an app on abap NW 7.02 and copy&paste it to an abap steampunk 2303 should work instantly and without additional configuration. Having this idea in mind, I checked the possibilities:
- openui5 - would be the easiest way, but they don’t support all UI5 controls. I added on friday charts, they are for example not a part of Openui5. For me this controls are very basic, they are used in nearly every Fiori app, therefore I can not skip them and therefore I can not use openui5
- 'resources/sap-ui-core-js' - every user has to create the icf service by themself, therefore the relativ part is different for every user, so this will end in an abapGit installation which is per default not runnning. Ui5 experts can fix this quickly, but normal ABAPers (which are the target group) have very often no idea about this, feels therefore not like the right way and a too complicated
- UI5 cloud CDN - it is the only working way for ABAP steampunk and trial user and abap2UI5 should work out of the box on an abap trial stack (also my dev environment at the moment, no abap dev edition anymore) In addition it supports the SAP cloud first strategy, every ABAP app which uses CDN is ready to migrate to ABAP PaaS. So I am a bit confused why the rules are so restricted here
- last point: my normal expectation would be that every company which want to use this productive will go to the Z2UI5_HTTP_HANDLER and set his favorite browser title, repository and theme before transporting it to P, so there will be no problem anyway, but we will see…

Long story short, I have no good idea how to solve this at the moment 🤨 but the project is very young, maybe it is better to wait a little bit and see how people are using it before making quick changes here.

I adjusted the title of the issue and will let it open. If you have more ideas how to solve this in an elegant way, feel free to let me know!

**Comment by marianfoo** [comment]
> Long story short, I have no good idea how to solve this at the moment 🤨 it seems that there is no working way which makes the project out-of-the-Box compatible to steampunk abap and on-prem abap with respecting sap license 🤷‍♂️

If that is your requirement, then I would suggest OpenUI5 as default, which would be overridable with other configurations (e.g. to use charts).

I think you yourself have the goal that it can/should be used easily in productive scenarios. But for this it must also stand on a solid and legal basis.

And yes, the theme is of course also dependent on the version. horizon is not available in the 1.71 version.

**Comment by mbtools** [comment]
In general, it's a bad idea to load resources for on-prem apps from the Internet. Many ABAP systems have no internet connection so access is only inside the firewall. A fixed external resource also won't work between dev, qa, and prod when you want to code against a new version. 

You really want to server any required JS (or other frontend resources) from the ABAP mime repo. The alternative is something like abapGit uses (W3MIME or a custom table). 

Cloud - esp public - is different (no mime, internet access is given). But still dev needs to be able to run a different resource version than prod. 

You can't ship UI5 with your repo so it must be configurable. I agree that it should be via code. Producing A2ui5 apps that run on cloud and on-prem and can be deployed without additional steps is a dream that's unfortunately not possible in the current SAP ecosystem.

**Comment by oblomov-dev** [comment]
app configuration has now an extra issue: https://github.com/oblomov-dev/abap2UI5/issues/129

**Comment by oblomov-dev** [comment]
"I think you yourself have the goal that it can/should be used easily in productive scenarios. But for this it must also stand on a solid and legal basis."
@marianfoo yes the longer i think about it, the more i also think openui5 is the right way as a default. i need to check the demos and the z2ui5_if_view for compatibility before i switch, but i hope there are no greater problems. maybe i can split the demo section into an openui5 section which is running immediately and an ui5 section (where you have to adjust the ui5 path before). just need to find a way that no normal abap programmer is getting too confused

"You can't ship UI5 with your repo so it must be configurable. I agree that it should be via code. Producing A2ui5 apps that run on cloud and on-prem and can be deployed without additional steps is a dream that's unfortunately not possible in the current SAP ecosystem."
@mbtools thanks for the info. changed the configuration today and made a new issue to find a good place for that: https://github.com/oblomov-dev/abap2UI5/issues/129

**Comment by marianfoo** [comment]
>  just need to find a way that no normal abap programmer is getting too confused


I think you underestimate the average ABAP developer who uses a package from GitHub and imported it with abapGit. :)

If you mark it clearly in the documentation as "Getting Started" or "Prerequisites" then it should be clear.
Of course, a clear error message would be best as well.

**Comment by oblomov-dev** [comment]
> I think you underestimate the average ABAP developer who uses a package from GitHub and imported it with abapGit. :)
haha yes, maybe I spend to many time with explaining the difference between fiori and ui5 or finding the right ui5 path in customer systems in my life 😅 you are right, abapGit is a good preselection 💪

**Comment by oblomov-dev** [comment]
I checked all demos and code now, so far only visualization (sap.suite.ui.microchart controls) are used which are not a part of openUI5. I splitted the demo section:
<img width="500" alt="image" src="https://user-images.githubusercontent.com/102328295/229784446-0f5ace4e-34f7-425f-ab24-531c7847dc0e.png">

Per default the openUI5 CDN is now used:
<img width="400" alt="image" src="https://user-images.githubusercontent.com/102328295/229784254-4480a9bb-b444-4e89-97b8-75d4e3561708.png">

Tried to sum up the knowledge of this issue in this documentation, that it is available for everyone:
https://github.com/oblomov-dev/abap2UI5/wiki/UI5-Library-and-Bootstrapping

I will close the issue now. If something in the license terms is changing or someone of you finds out more about the topic, feel free to reopen it or make a new issue!

Thanks for the input!
